### PR TITLE
chore: GHA to smoke test against last released exemplar

### PR DESCRIPTION
**Pull request by safeer**
This adds a GitHub Action that smoke tests branches against the exemplar in the latest tagged release.

It will fetch the latest tagged FTL release from dockerhub, checkout the repo at that tag, then deploy that version to a local kube cluster. Testing this tagged version against the current exemplar is disabled (chicken-and-egg since test change are needed in the tagged release). Finally, it will upgrade the local kube cluster to that in the current branch and run the exemplar smoketest to verify the upgrade path was successful.

### CODE REVIEW

**Review by stuartwdouglas — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
I think it should actually be possible to run the upgrade tests on github actions. 

Basically you could do something like:

- checkout the last release
- Replace the images in deployments with the actual release images
- run the smoke test which will give you a cluster setup with the old version of both FTL and exemplar
- Build the new FTL and deploy it to the cluster, and run the smoke test (to verify old deployment works with FTL)
- Build and deploy the new version of the exemplar, and run the smoke test again

This would require a little bit of messing around with the build files but I think it should be doable.

File: .github/workflows/ci.yml
Comment (alecthomas) [inline .github/workflows/ci.yml:410]: Isn't the repo still going to be on LATEST_RELEASE? Where's the upgrade occurring?

**Review by alecthomas — COMMENTED** [review]
This shouldn't be in ci.yml, it should be in its own workflow and should block the release.

**Comment by stuartwdouglas** [comment]
> This shouldn't be in ci.yml, it should be in its own workflow and should block the release.

IMHO we still need a way to run this on pull requests, if you are doing something that may affect compatibility it should be possible to test this with 'run-all'.

File: .github/workflows/ci.yml
Comment (stuartwdouglas) [reply .github/workflows/ci.yml:410]: This doesn't actually test the upgrade yet, it is just testing that the old version of the exemplar still works with the current FTL.
While this is a start I think we should enhance it as per my comment above. I don't think we need an external cluster to test this, we should be able to adapt the existing kube tests to perform an upgrade.

**Comment by alecthomas** [comment]
> IMHO we still need a way to run this on pull requests, if you are doing something that may affect compatibility it should be possible to test this with 'run-all'. 

That's fine, but it should also run on main merge and/or during release, which means it can't be in ci.yml

File: .github/workflows/ci.yml
Comment (safeer) [reply .github/workflows/ci.yml:410]: I hit some kube snags trying to do this, but I'll dig in some more. Also, it seems as though the kube integration test support is for Go only atm

**Comment by safeer** [comment]
After syncing with @stuartwdouglas, the current plan is to:

1. checkout the latest tagged rlease
1. update ./overlay/smoketest/kustomization.yml to use the same tagged docker hub images
1. deployment Justfile to full-deploy a cluster
1. run the smoke tests on the cluster
1. checkout HEAD of the current branch
1. update ./overlay/smoketest/kustomization.yml to use latest image
1. deployments Justfile to build and apply the latest image to the cluster 
1. run the smoke tests on the cluster

Will require some work to switch out the kustomization.yml tag, and to allow `in.WithKubernetes` to conditionally do a `full-deploy`, since we want to skip the build portion when pushing tagged docker hub releases.

**Comment by safeer** [comment]
The upgrade portion is in and working; we deploy the latest tagged dockerhub image to the cluster, build the version from HEAD and upgrade the cluster, then run the test suite.

I'm skipping running the tagged release tests on the current branch for now, since I had to make changes to the integration test suite, so it's a chicken-and-egg. I'll follow this up when the change is in a tagged release.

I'm also only running the contrived Go test. Next step is to add the JDK to test the full exemplar.

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:62]: Everything is handled by kube, we should not need docker compose or initdb

Comment (safeer) [reply .github/workflows/smoketest.yml:62]: Thanks, yeah gonna do some cleanup here

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:2]: I think we actually want this to be on push, as we are not using merge groups any more.

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:11]: I think we need to uncomment this before merging.

File: .github/workflows/ci.yml
Comment (stuartwdouglas) [inline .github/workflows/ci.yml:378]: I think this should be removed from ci.yaml, as it is duplicated in smoketest.yaml, which also runs on PRs?

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:22]: At some point (not now) we are going to need to think about testing older releases. The 'latest release' is often just a few commits ago.

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:50]: This approach will break if the smoke test adds new features that are in the new version but not the old version, as you are testing the new code again old FTL.

This is fine for now but I think but will need some thought going forward.

File: .github/workflows/smoketest.yml
Comment (stuartwdouglas) [inline .github/workflows/smoketest.yml:72]: `wait-for-kube` is already in the harness, if you leave it off here it will speed up the test.

**Comment by safeer** [comment]
Ingress is working, but looks like PubSub is also borked in the kube environment

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**
